src/TestExplorer/TestRunner.ts

```diff
@@ -726,49 +726,58 @@ class TestRunnerTestRunState implements iTestRunState {
         lineNumber: number;
         complete: boolean;
     };
+    private startTimes: Map<number, number | undefined> = new Map();
+    private issues: Map<number, vscode.TestMessage[]> = new Map();
 
     getTestItemIndex(id: string, filename?: string): number {
         return this.testItemFinder.getIndex(id, filename);
     }
 
     // set test item to be started
-    started(index: number): void {
+    started(index: number, startTime?: number): void {
         this.testRun.started(this.testItemFinder.testItems[index]);
         this.currentTestItem = this.testItemFinder.testItems[index];
+        this.startTimes.set(index, startTime);
     }
 
     // set test item to have passed
-    passed(index: number, duration: number): void {
-        this.testRun.passed(this.testItemFinder.testItems[index], duration * 1000);
-        this.testItemFinder.testItems.splice(index, 1);
+    completed(index: number, durationOrEndTime: number): void {
+        const test = this.testItemFinder.testItems[index];
+        const startTime = this.startTimes.get(index);
+        const duration =
+            (startTime !== undefined ? durationOrEndTime - startTime : durationOrEndTime) * 1000;
+        const issues = this.issues.get(index) ?? [];
+
+        if (issues.length > 0) {
+            this.testRun.failed(test, issues, duration);
+        } else {
+            this.testRun.passed(test, duration);
+        }
+
         this.lastTestItem = this.currentTestItem;
         this.currentTestItem = undefined;
     }
 
-    // set test item to be failed
-    failed(index: number, message: string, location?: { file: string; line: number }): void {
+    recordIssue(
+        index: number,
+        message: string | vscode.MarkdownString,
```

Comment:
Any reason this is also a `MarkdownString`?

Reply:
This message gets passed into a [vscode.TestMessage](https://github.com/microsoft/vscode/blob/main/src/vs/workbench/api/common/extHostTypes.ts#L4046), which accepts `string | MarkdownSting`. 

I'd like swift-testing to produce a markdown string of the issue description, which I noted here https://github.com/apple/swift-testing/issues/368

Reply:
Ok